migration: Add CASCADE option to DROP command

mpotra opened the pull request:
This PR provides the ability to add the `CASCADE` option to `DROP` commands in migrations, on tables and indexes.
The default option on all DROP commands is always `RESTRICT` (implied default)

PostgreSQL documentation: 
- https://www.postgresql.org/docs/current/sql-droptable.html
- https://www.postgresql.org/docs/current/sql-dropindex.html

MySQL documentation:
- https://dev.mysql.com/doc/refman/8.0/en/drop-table.html
- no support on `DROP INDEX` for `CASCADE` or `RESTRICT`

TDS/MSSQL:
- Does not support `CASCADE` | `RESTRICT` params for any `DROP` command.

Migration `drop` commands [will now support](https://github.com/elixir-ecto/ecto_sql/pull/298/files#diff-da3c3b102234918cbfd1348495384893ed67e47302c350051d1d78eeb1ecf6fc) `mode :cascade` option

Examples:
`drop table("users"), mode: :cascade` -> `DROP TABLE "users" CASCADE`
`drop_if_exists table("users"), mode: :cascade` -> `DROP TABLE IF EXISTS "users" CASCADE`
`drop index("posts", [:name]), mode: :cascade` -> `DROP INDEX posts_name_index CASCADE`
`drop_if_exists index("posts", [:name]), mode: :cascade` -> `DROP INDEX IF EXISTS posts_name_index CASCADE`

josevalim commented:
Sorry, this fell through the cracks. First of all, thanks for the PR!

I have two suggestions:

1. Instead of cascade: true, let's have `mode: :cascade | :restrict`. In this case, we will change the drop tuple to always be, `{:drop, index | table, mode}`. This should simplify some of the code.

2. Please check how it should behave with TDS :)

josevalim commented:
Ping!

mpotra commented:
Sorry, completely forgot about this PR. Will address the changes in a few days 🙏

mpotra commented:
After a while, I've switched to using the `:mode` option and adjusted tests for this as well.

TDS does not support cascaded drops, so it is ignored.

josevalim commented:
Thank you! One quick note, if TDS does not support it, it should raise if you are not using the default value. :)

mpotra commented:
> Thank you! One quick note, if TDS does not support it, it should raise if you are not using the default value. :)

Makes perfect sense, to raise if `:cascade` mode is used when not supported in TDS; thank you!

I've pushed the commit with the changes as per above, but seems that the last test failed due to some mssql connection issue when starting the mssql server.
Could we rerun the last test only?

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
